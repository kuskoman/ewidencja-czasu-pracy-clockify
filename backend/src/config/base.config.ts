import { ConfigType, registerAs } from '@nestjs/config';

export const BASE_CONFIG_KEY = 'BASE_CONFIG';

export const baseConfig = registerAs(BASE_CONFIG_KEY, () => ({
  port: parseInt(`${process.env.PORT || 3000}`),
}));

export type BaseConfig = ConfigType<typeof baseConfig>;
