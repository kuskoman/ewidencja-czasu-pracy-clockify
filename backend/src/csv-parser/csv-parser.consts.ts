export const CSV_REPORT_DATE_FORMAT = 'DD/MM/YYYY';
