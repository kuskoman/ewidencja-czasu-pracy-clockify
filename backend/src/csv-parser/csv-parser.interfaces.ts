export type CsvReport = [string, unknown, string][];
export type SerializedReport = Array<{ date: Date; time: number }>;
