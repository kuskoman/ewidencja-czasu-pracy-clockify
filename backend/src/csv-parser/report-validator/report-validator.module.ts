import { Module } from '@nestjs/common';
import { ReportValidatorService } from './report-validator.service';

@Module({
  providers: [ReportValidatorService],
  exports: [ReportValidatorService],
})
export class ReportValidatorModule {}
