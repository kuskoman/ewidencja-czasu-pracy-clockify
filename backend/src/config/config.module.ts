import { Module } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { baseConfig } from './base.config';
import { csvConfig } from './csv.config';
import { swaggerConfig } from './swagger.config';

@Module({
  imports: [
    ConfigModule.forRoot({
      cache: true,
      load: [baseConfig, swaggerConfig, csvConfig],
      isGlobal: true,
    }),
  ],
})
export class AppConfigModule {}
