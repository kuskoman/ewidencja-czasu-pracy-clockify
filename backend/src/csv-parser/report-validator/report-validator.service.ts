import { Injectable, UnprocessableEntityException } from '@nestjs/common';
import { CSV_REPORT_DATE_FORMAT } from '../csv-parser.consts';
import { CsvReport } from '../csv-parser.interfaces';
import { utc } from 'moment';

@Injectable()
export class ReportValidatorService {
  public validateCsvReport(report: unknown[][]): report is CsvReport {
    if (!Array.isArray(report)) {
      throw new UnprocessableEntityException('Report is not an array');
    }

    if (report.length < 1) {
      throw new UnprocessableEntityException("Report can't be empty");
    }

    const firstElement = report[0][0];
    if (firstElement === 'Date') {
      const errMsg =
        'Expected to receive raw csv data, received file with headers';
      throw new UnprocessableEntityException(errMsg);
    }

    for (const row of report) {
      if (row.length != 3) {
        const errMsg = `Report row has invalid number of columns: ${row.length}`;
        throw new UnprocessableEntityException(errMsg);
      }

      const [date, , decimalTime] = row;
      if (typeof date !== 'string' || typeof decimalTime !== 'string') {
        throw new UnprocessableEntityException('Received invalid type of data');
      }

      if (!this.isValidDate(date)) {
        const errMsg = `Expected valid date, received ${date}`;
        throw new UnprocessableEntityException(errMsg);
      }

      if (!this.isValidFloat(decimalTime)) {
        const errMsg = `Expected valid decimalTime, received ${decimalTime}`;
        throw new UnprocessableEntityException(errMsg);
      }
    }

    return true;
  }

  private isValidDate(dateString: string) {
    const parsedDate = utc(dateString, CSV_REPORT_DATE_FORMAT);
    return parsedDate.isValid();
  }

  private isValidFloat(numberString: string) {
    return !Number.isNaN(Number(numberString));
  }
}
