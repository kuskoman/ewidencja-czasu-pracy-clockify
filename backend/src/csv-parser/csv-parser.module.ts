import { Module } from '@nestjs/common';
import { CsvParserService } from './csv-parser.service';
import { ReportValidatorModule } from './report-validator/report-validator.module';

@Module({
  providers: [CsvParserService],
  imports: [ReportValidatorModule],
  exports: [CsvParserService],
})
export class CsvParserModule {}
