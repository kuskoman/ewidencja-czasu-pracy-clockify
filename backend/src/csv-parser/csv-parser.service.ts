import {
  Inject,
  Injectable,
  UnprocessableEntityException,
} from '@nestjs/common';
import { CsvConfig, csvConfig } from '../config/csv.config';
import { parse as parseCsv } from 'csv';
import { ReportValidatorService } from './report-validator/report-validator.service';
import { CsvReport, SerializedReport } from './csv-parser.interfaces';
import * as moment from 'moment';
import { CSV_REPORT_DATE_FORMAT } from './csv-parser.consts';

@Injectable()
export class CsvParserService {
  constructor(
    @Inject(csvConfig.KEY) private readonly config: CsvConfig,
    private readonly reportValidator: ReportValidatorService,
  ) {}

  public async parseCsvReport(csv: string | Buffer) {
    const parser = parseCsv(csv, {
      delimiter: this.config.delimeter,
      fromLine: 2,
    });

    const parsedData = await this.parserToPromise(parser);
    if (!this.reportValidator.validateCsvReport(parsedData)) {
      throw new UnprocessableEntityException('CSV validation failed');
    }

    return this.serializeCsvReport(parsedData);
  }

  private serializeCsvReport(csvReport: CsvReport): SerializedReport {
    return csvReport.map(this.serializeCsvRow);
  }

  private serializeCsvRow(row: CsvReport[number]) {
    const [dateString, , decimalTime] = row;

    return {
      date: moment(dateString, CSV_REPORT_DATE_FORMAT).toDate(),
      time: Number(decimalTime),
    };
  }

  private parserToPromise(parser: Parser): Promise<unknown[][]> {
    const parsedData: unknown[][] = [];
    return new Promise((res, rej) => {
      parser
        .on('data', (row) => {
          parsedData.push(row);
        })
        .on('error', rej)
        .on('end', () => {
          res(parsedData);
        });
    });
  }
}

type Parser = ReturnType<typeof parseCsv>;
