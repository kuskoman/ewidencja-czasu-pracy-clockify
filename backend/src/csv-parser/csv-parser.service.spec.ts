import { Test, TestingModule } from '@nestjs/testing';
import { readFileSync } from 'fs';
import { join } from 'path';
import { csvConfig, CsvConfig } from '../config/csv.config';
import { CsvParserService } from './csv-parser.service';
import { ReportValidatorModule } from './report-validator/report-validator.module';

describe(CsvParserService.name, () => {
  let service: CsvParserService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        CsvParserService,
        { provide: csvConfig.KEY, useValue: csvConfigMock },
      ],
      imports: [ReportValidatorModule],
    }).compile();

    service = module.get<CsvParserService>(CsvParserService);
  });

  it('should properly parse valid csv', async () => {
    const validReportFixture = loadFixture('example-report.csv');
    const parsedCsv = await service.parseCsvReport(validReportFixture);
    expect(parsedCsv).toMatchSnapshot();
  });
});

const csvConfigMock: CsvConfig = {
  delimeter: ',',
};

const loadFixture = (name: string): string => {
  const filePath = join(__dirname, 'fixtures', name);
  return readFileSync(filePath, 'utf-8');
};
