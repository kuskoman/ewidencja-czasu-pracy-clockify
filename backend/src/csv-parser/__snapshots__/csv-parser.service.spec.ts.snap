// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CsvParserService should properly parse valid csv 1`] = `
[
  {
    "date": 2022-10-02T22:00:00.000Z,
    "time": 7.933333333333334,
  },
  {
    "date": 2022-10-03T22:00:00.000Z,
    "time": 7.383333333333334,
  },
  {
    "date": 2022-10-04T22:00:00.000Z,
    "time": 7.916666666666667,
  },
  {
    "date": 2022-10-05T22:00:00.000Z,
    "time": 8.8175,
  },
  {
    "date": 2022-10-06T22:00:00.000Z,
    "time": 4.6963888888888885,
  },
  {
    "date": 2022-10-09T22:00:00.000Z,
    "time": 8.655,
  },
  {
    "date": 2022-10-10T22:00:00.000Z,
    "time": 5.633055555555556,
  },
  {
    "date": 2022-10-11T22:00:00.000Z,
    "time": 8.386111111111111,
  },
  {
    "date": 2022-10-12T22:00:00.000Z,
    "time": 7.6675,
  },
  {
    "date": 2022-10-13T22:00:00.000Z,
    "time": 7.371111111111111,
  },
  {
    "date": 2022-10-16T22:00:00.000Z,
    "time": 2,
  },
  {
    "date": 2022-10-17T22:00:00.000Z,
    "time": 9.56638888888889,
  },
  {
    "date": 2022-10-18T22:00:00.000Z,
    "time": 8.058333333333334,
  },
  {
    "date": 2022-10-19T22:00:00.000Z,
    "time": 8.032777777777778,
  },
  {
    "date": 2022-10-20T22:00:00.000Z,
    "time": 6.892222222222222,
  },
  {
    "date": 2022-10-21T22:00:00.000Z,
    "time": 1.1688888888888889,
  },
  {
    "date": 2022-10-23T22:00:00.000Z,
    "time": 8.724444444444444,
  },
  {
    "date": 2022-10-24T22:00:00.000Z,
    "time": 8.165,
  },
  {
    "date": 2022-10-25T22:00:00.000Z,
    "time": 7.833333333333333,
  },
  {
    "date": 2022-10-26T22:00:00.000Z,
    "time": 9.016111111111112,
  },
  {
    "date": 2022-10-27T22:00:00.000Z,
    "time": 10.085555555555555,
  },
  {
    "date": 2022-10-30T23:00:00.000Z,
    "time": 7.8438888888888885,
  },
]
`;
